### PR TITLE
Don't override relatedImages in base CSV. #5763

### LINKED ISSUE #5763: Don't override relatedImages in config CSV.
## Bug Report
Looks like https://github.com/operator-framework/operator-sdk/issues/5000 has reappeared after related image discovery features were added. This has caused us to change how we specify relatedImages in our operator https://github.com/openshift/oadp-operator/pull/694

Seems to be sideeffect from related image discovery feature. Not sure which was the first PR for that.

Related PRs: https://github.com/operator-framework/operator-sdk/pull/5567

#### What did you do?
operatorsdk generate bundle.

#### What did you expect to see?
relatedImages in config CSV preserved

#### What did you see instead? Under which circumstances?
relatedImages in config CSV removed from generated bundle

#### Environment

**Operator type:**

/language go

`$ operator-sdk version`

operator-sdk version: "v1.20.0", commit: "deb3531ae20a5805b7ee30b71f13792b80bd49b1", kubernetes version: "v1.23", go version: "go1.18.1", GOOS: "darwin", GOARCH: "amd64"
`$ go version` (if language is Go)

go version go1.18.1 darwin/amd64
`$ kubectl version`

#### Possible Solution
Don't override but append to existing relatedImages.

### PR DESCRIPTION
**Pull request by kaovilai**
Closes #5763, #5000

**Description of the change:**
Regression: relatedImages in base CSV were being overridden.

**Motivation for the change:**
Do not want relatedImages overridden from baseCSV.

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [x] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [ ] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

### CODE REVIEW

File: internal/generate/clusterserviceversion/clusterserviceversion.go
Comment (joelanford) [inline internal/generate/clusterserviceversion/clusterserviceversion.go:170]: We might want to de-duplicate here.

File: changelog/fragments/do-not-override-base-csv-relatedImages.yaml
Comment (joelanford) [inline changelog/fragments/do-not-override-base-csv-relatedImages.yaml:37]: Remove this since there's no migration

File: changelog/fragments/do-not-override-base-csv-relatedImages.yaml
Comment (joelanford) [inline changelog/fragments/do-not-override-base-csv-relatedImages.yaml:5]:
```suggestion
      Fixes an issue that caused hardcoded values in the base CSV's `spec.relatedImages` field to be overwritten during bundle generation.
```

**Review by dymurray — APPROVED** [review]
Worked for me

File: internal/generate/clusterserviceversion/clusterserviceversion.go
Comment (joelanford) [inline internal/generate/clusterserviceversion/clusterserviceversion.go:181]: Sorry - I should have been more specific in my previous comment about de-duplicating.

Can we key this by the image field rather than the name? That's ultimately what matters. The relatedImage name is just informational and is not actually used by any of the mirroring tooling as far as I know.

We already have [some code](https://github.com/operator-framework/operator-sdk/blob/f62caa6579d14883155b91e241a98272d7a9c161/internal/cmd/operator-sdk/generate/internal/relatedimages.go#L125-L142) that does this for the `g.RelatedImages`. I wonder if we can combine `base.Spec.RelatedImages` earlier in the process such that that code picks them up and uses the same de-dup and warn logic?

In the case where there are multiple images with different names, I think @ryantking

Comment (joelanford) [reply internal/generate/clusterserviceversion/clusterserviceversion.go:181]: cc @ryantking

**Comment by kaovilai** [comment]
@ryantking What's the point of appending `containerRef` on name collision if you are doing it on all collisions anyway? To make it obvious which are colliding?
https://github.com/operator-framework/operator-sdk/blob/fb15b62e9c0defee208086ccab79112c68843454/internal/cmd/operator-sdk/generate/internal/relatedimages.go#L120-L123

ie. I've just tried in my change putting this in base csv.
```
  relatedImages:
  - image: quay.io/konveyor/velero-restic-restore-helper:latest
    name: velero_restic_restore_helper
  - image: quay.io/konveyor/velero-restic-restore-helpser:latest
    name: velero_restic_restore_helper
```
And got back this generated csv.
```
  relatedImages:
    - image: quay.io/konveyor/velero:latest
      name: velero
    - image: quay.io/konveyor/velero-restic-restore-helper:latest
      name: baseCSV-velero_restic_restore_helper
    - image: quay.io/konveyor/velero-restic-restore-helpser:latest
      name: baseCSV-velero_restic_restore_helper
```

**Comment by kaovilai** [comment]
Result is still two relatedImages with `      name: baseCSV-velero_restic_restore_helper`

File: internal/generate/clusterserviceversion/clusterserviceversion.go
Comment (kaovilai) [reply internal/generate/clusterserviceversion/clusterserviceversion.go:181]: As far as I can tell, the warning only exists for same ref but different name which I guess is all we care about.

Comment (kaovilai) [reply internal/generate/clusterserviceversion/clusterserviceversion.go]: resolved via https://github.com/operator-framework/operator-sdk/commit/49622d5c939e6e22c5d54c3878697db811473116

**Comment by kaovilai** [comment]
`warning: multiple related images with the same image ref, "openshift_plugin", and different names found.The image will only be listed once with an empty name.It is recmmended to either remove the duplicate or use the exact same name. `

Is it a good idea to change image name to empty? Is this to make a user notice what they need to fix?
Is it okay to leave as is or will this cause failures with mirroring later on?
```
    - image: quay.io/konveyor/openshift-velero-plugin:latest
      name: ""
```

**Comment by kaovilai** [comment]
> The name is only prefixed with there are two images with the same name, but different images.

This does not fix the collision still. Is it just a notification mechanism?

**Comment by kaovilai** [comment]
I think a more obvious is to name something in all caps.. like OPERATOR_SDK_NAME_CONFLICT_originalname

**Comment by kaovilai** [comment]
Sounds like both "" and containerRef appending are notification mechanism.. but not a very consistent one.. since one adds to the name, and one removes the name.

**Comment by ryantking** [comment]
@kaovilai Okay I think I have a better idea of what you're pointing out now. The collision-resolution logic might not correctly resolve colissions when one or both related images are defined in the base CSV.

The logic operates under the assumption that collisions will occur across different containers and deployments. We made this assumption originally since a colission within a container would require two environment variables with the same key in a single container environment, which I believe Kubernetes would not allow. The edge case introduced here makes it so two relatedImages can have the same name, different image, but the same containerRef that war previously assumed to be different if at least one was hardcoded into the base CSV.

I'm not entirely sure what the best resolution is since the relatedImage collector logic assumes we have knowledge of the source container, which we don't for hardcoded relatd images. To get more understanding of the usecase, what is the reasoning behind defining relatedImages in the base CSV spec instead of as environment variables in one or more containers, which will then get picked up when building the output CSV?

> Sounds like both "" and containerRef appending are notification mechanism.. but not a very consistent one.. since one adds to the name, and one removes the name.

Neither were defined to be notification mechanisms. Prepending the containerRef was at the time a way to actually resolve conflicts since we didn't have this edge case yet, and empty string for image collisions is because the program cannot determine which name to use if the same image has two different names. The warning that gets printed is the notification mechanism. Maybe we should also warn on name collisions, but that's an aside.

**Comment by kaovilai** [comment]
@ryantking same usecase as previously stated in https://github.com/operator-framework/operator-sdk/issues/5000 which links to https://docs.openshift.com/container-platform/4.7/operators/operator_sdk/osdk-generating-csvs.html#olm-enabling-operator-for-restricted-network_osdk-generating-csvs that says to update
> Update spec.relatedImages

directly. They made no mention of RELATED_IMAGE_ in deployment.

This was how it was always done before your relatedImage discovery feature.

**Comment by kaovilai** [comment]
The specified relatedImage are potential images required in disconnected environments of supported [plugins](https://velero.io/plugins/) that user may choose to enable after.

**Comment by ryantking** [comment]
@kaovilai The correct procedure is defined in the [latest documentation](https://docs.openshift.com/container-platform/4.10/operators/operator_sdk/osdk-generating-csvs.html#olm-enabling-operator-for-restricted-network_osdk-generating-csvs). It instructs you to update the environment in `manager.yaml` and explains how to use them in the operator code.

Let me know if that solves your usecase. We still probably want to have a discussion on what to do with any relatedImages in the base CSV, but before that we need to establish a usecase where they must be defined in the base CSV and cannot be defined as environment variables the way we officially support.

If the official method works for you then I'm going to hold this PR and have a longer discussion with the team on how to handle existing relatedImage values in the base CSV to decide how to move forward.

**Comment by kaovilai** [comment]
Resolving our team's relatedImages would definitely require more than dev team code changes since our release team have scripts which expects certain relatedImage names today. Per https://github.com/openshift/oadp-operator/pull/694 there are sideeffects from migrating to this new "supported" relatedImage specification which include changing relatedImage name to `-` instead of `_`

**Comment by ryantking** [comment]
@kaovilai Okay so it looks like the related image dicovery feature introduced a breaking change that I didn't notice (and therefore document correctly). Short term will likely be a documentation update and recommending that users define their related images in `manager.yaml` as the latest docs say. I will investigate the changes required to introduce backwards compatibility with directly defining relatedImages in the base CSV. Ideally I'd like to "unbreak" this change, but losing the container/deployment info is going to make the dedup logic a little trickier.

Since this PR cannot move forward with the existing de-dup logic, I am holding this for now. If you'd still like to move this PR forward then I recommend joining our SDK community meeting Wednesday at 2PM EST to discuss it in more detail. Join the [Google Group](https://groups.google.com/g/operator-framework) to get the invites on your google calendar if you don't have them already. If you can't make that meeting and still want to move this PR forward then contact me privately via email or Slack and I can get something together with us and anyone else with context to figure out a solution for the de-dup logic. 

If you're fine updating to the official `manager.yaml` method and don't want to personally drive this fix then I recommend closing the PR.

/hold

**Comment by kaovilai** [comment]
I'll try to make the meeting on Wednesday. Sounds like it maybe a good learning experience for me :)

**Comment by kaovilai** [comment]
IIUC you're referring to 
> * Backlog Grooming Meeting: Wednesdays (every 3 weeks) at 11:00 Western Time. [Convert to your timezone](http://www.thetimezoneconverter.com/?t=11:00&tz=PT%20%28Western%20Time%29).
    * [Meeting Link][meeting-link]: Password: 77777
    * [Agenda](https://docs.google.com/document/d/1ujWb-rSJ4JWeHLVxK0WS5ZuSJgeESG42MDeYjSl9Q6U/edit)
    * [Meeting recordings](https://www.youtube.com/playlist?list=PLEcO8aSeUjeXxkVh27ExScB_wEmhBLY7g).

https://github.com/operator-framework/community/blob/06ba4337c79560448242eca36758b6db3f75927b/README.md?plain=1#L67-L70

**Comment by kaovilai** [comment]
> In order to ensure total backwards compatiblity, the base CSV related images need to be preserved entirely so they should take precedence over discovered related images when resolving name conflicts.

Summarized dedupe precedence that will be implemented.
https://github.com/operator-framework/operator-sdk/issues/5763#issuecomment-1128358142

**Comment by kaovilai** [comment]
Additionally add warnings.
> issue a warning about the statically defined related images. If we want to be more vague about why you shouldn't do that, we could just say that specifying static related images is deprecated, that you should start using RELATED_IMAGE_ envs, and we'll remove persistence of static related images is OSDK v2.

File: internal/cmd/operator-sdk/generate/internal/relatedimages.go
Comment (kaovilai) [inline internal/cmd/operator-sdk/generate/internal/relatedimages.go:156]: If a related image with the same name/image was discovered, do nothing. It's already there.

File: internal/cmd/operator-sdk/generate/internal/relatedimages.go
Comment (kaovilai) [inline internal/cmd/operator-sdk/generate/internal/relatedimages.go:145]: If a related image with the same name was discovered, prefix the instances after first discovered name with its container name, leave the base CSV definition as is (if they don't have more than one relatedImage with the same name) and add it to the final CSV.

Comment (kaovilai) [reply internal/cmd/operator-sdk/generate/internal/relatedimages.go:145]: First discovered name will be coming from baseCSV first.

File: internal/cmd/operator-sdk/generate/internal/relatedimages.go
Comment (kaovilai) [inline internal/cmd/operator-sdk/generate/internal/relatedimages.go:133]: If a related image with the same image was discovered, just add it in.

File: internal/cmd/operator-sdk/generate/internal/relatedimages.go
Comment (kaovilai) [inline internal/cmd/operator-sdk/generate/internal/relatedimages.go:160]: Warn the user an image is used twice.

File: internal/cmd/operator-sdk/generate/internal/relatedimages.go
Comment (kaovilai) [inline internal/cmd/operator-sdk/generate/internal/relatedimages.go:95]: per https://github.com/operator-framework/operator-sdk/issues/5763#issuecomment-1129185789

>  issue a warning about the statically defined related images. If we want to be more vague about why you shouldn't do that, we could just say that specifying static related images is deprecated, that you should start using RELATED_IMAGE_ envs, and we'll remove persistence of static related images is OSDK v2.

**Review by kaovilai — COMMENTED** [review]
comment on recent changes.

File: .vscode/settings.json
Comment (ryantking) [inline .vscode/settings.json:3]: Delete, you should probably add `.vscode` to `$HOME/.gitignore_global`

File: internal/cmd/operator-sdk/generate/internal/relatedimages.go
Comment (ryantking) [inline internal/cmd/operator-sdk/generate/internal/relatedimages.go:29]: I don't like passing in the `packageName` here then using it to detect the base manifest. The function you added below could potentiall produce a false positive since it does not enforce the base CSV being found in the `bases` directory [as sdk does in the logic that calls this function][https://github.com/operator-framework/operator-sdk/blob/master/internal/cmd/operator-sdk/generate/bundle/bundle.go#L183]. We should either make the manifest collector aware of the base CSV so that way you can call `manifestCol.GetBaseClusterServiceVersion()` or pass in `overrides map[string]string` such that its a map of all relatedImages found in the root of the generate command.

File: internal/cmd/operator-sdk/generate/internal/relatedimages.go
Comment (ryantking) [inline internal/cmd/operator-sdk/generate/internal/relatedimages.go:92]: We shouldn't be calling this function at all if there's no base CSV. The caller should own checking if it is present or not. Also, you don't need to nil check the related images since they are in a slice you can safely iterate over a `nil` slice so long as its typed:

```go
x := len(nil) // will panic
y := len([]int(nil)) // won't panic
```

File: internal/cmd/operator-sdk/generate/internal/relatedimages.go
Comment (ryantking) [inline internal/cmd/operator-sdk/generate/internal/relatedimages.go:94]: I don't think we want to use `c.collect` here. That function has a bunch of logic to handle collisions, which we don't want applied to base CSVs. Base CSV related Images should just be moved over as is. I would think what we want is the final step of collecting related images is going through all the base CSV related images and for each:

1. If a discovererd related image has the same name and image: do nothing, we already have that an equivalent related image in the output.
2. If a discovererd related image has the same name, but different image: prefix the discovered related image name, put the base CSV related image in the output as is.
3. If a discovered related image has the same image, but different name: put the base CSV related image in the output as is. We typically don't want the same image listed twice, but in this case we accept it in the name of backwards compatibility.

All should have a clear warning that this is a deprecated approach, and will be removed in v2.

File: internal/cmd/operator-sdk/generate/internal/relatedimages.go
Comment (ryantking) [inline internal/cmd/operator-sdk/generate/internal/relatedimages.go:108]: This is out of the responsibility of this package. All handling of pinned images is done in [this library][https://github.com/operator-framework/operator-manifest-tools/tree/main/pkg/image].

File: internal/cmd/operator-sdk/generate/internal/relatedimages.go
Comment (ryantking) [inline internal/cmd/operator-sdk/generate/internal/relatedimages.go:140]: This logic is subject to change based on my previous comments, but you should use `sets.String` from https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets instead of a map when you are just using it to track the existence of a unqiue set of strings.

**Review by ryantking — CHANGES_REQUESTED** [review]
Looks good, but I think we should totally decouple the base CSV logic from this package. This package should only know how to discover related images in the environment. [Right here][https://github.com/operator-framework/operator-sdk/blob/master/internal/cmd/operator-sdk/generate/bundle/bundle.go#L199] we should apply the base CSV related images as a final step so that way we ensure backwards compatibility and can preserve this logic.

Feel free to reach out on slack or email if you want to talk about this in more detail.

File: .vscode/settings.json
Comment (kaovilai) [reply .vscode/settings.json:3]: +1

**Comment by kaovilai** [comment]
just fyi that our workload are no longer affected by this issue. Happy to let others takeover.

**Pull request closed without merging**
